masonry/src/passes/update.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright 2024 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashSet;
+
+use cursor_icon::CursorIcon;
+use tracing::trace;
+
+use crate::passes::merge_state_up;
+use crate::render_root::{RenderRoot, RenderRootSignal};
+use crate::{LifeCycleCtx, StatusChange, Widget, WidgetId, WidgetState};
+
+fn get_id_path(root: &RenderRoot, widget_id: Option<WidgetId>) -> Vec<WidgetId> {
+    let Some(widget_id) = widget_id else {
+        return Vec::new();
+    };
+
+    root.widget_arena
+        .widget_states
+        .get_id_path(widget_id.to_raw())
+        .iter()
+        .map(|&id| WidgetId(id.try_into().unwrap()))
+        .collect()
+}
+
+// TODO - Replace LifecycleCtx with UpdateCtx
+pub(crate) fn run_targeted_update_pass(
+    root: &mut RenderRoot,
+    target: Option<WidgetId>,
+    pass_fn: impl FnMut(&mut dyn Widget, &mut LifeCycleCtx),
+) {
+    let mut pass_fn = pass_fn;
+
+    let mut target_widget_id = target;
+    while let Some(widget_id) = target_widget_id {
+        let parent_id = root.widget_arena.parent_of(widget_id);
+        let (widget_mut, state_mut) = root.widget_arena.get_pair_mut(widget_id);
+
+        let mut ctx = LifeCycleCtx {
+            global_state: &mut root.state,
+            widget_state: state_mut.item,
+            widget_state_children: state_mut.children,
+            widget_children: widget_mut.children,
+        };
+        pass_fn(widget_mut.item, &mut ctx);
+
+        merge_state_up(&mut root.widget_arena, widget_id);
+        target_widget_id = parent_id;
+    }
+}
+
+pub(crate) fn run_update_pointer_pass(root: &mut RenderRoot, root_state: &mut WidgetState) {
+    let pointer_pos = root.last_mouse_pos.map(|pos| (pos.x, pos.y).into());
+
+    // -- UPDATE HOVERED WIDGETS --
+
+    let hovered_widget = root.state.hovered_path.last().copied();
+    let mut next_hovered_widget = if let Some(pos) = pointer_pos {
+        // TODO - Apply scale?
+        root.get_root_widget()
+            .find_widget_at_pos(pos)
+            .map(|widget| widget.id())
+    } else {
+        None
+    };
```

Comment:
I wonder if the optimisation of "start from the previous hovered widget" is worthwhile.
i.e. special case the common case where the mouse stays within the same widget.

(Notice that we need to be careful to handle any children of that hovered widget, and any reparenting; i.e. this isn't really for this PR )

Reply:
> I wonder if the optimisation of "start from the previous hovered widget" is worthwhile.

I doubt it. For anything `O(Depth)` the overhead is already small enough that any optimization that makes the code more complex / adds hidden assumptions is a bad trade.